tools/scoring/score.py

```diff
@@ -568,28 +606,124 @@ def project(normal,other):
     other_different = different.loc['other']
     return other_different
 
+# Do not consider solver variants for determining if a division is
+# competitive.
+# Uses solver_str(solver) to get the base version of solver.
+def is_division_competitive(solvers):
+    return len(set([solver_str(x) for x in solvers])) > 1
+
+
+# Biggest Lead Ranking.
+#
+# Computes the new global ranking based on the distance between the winner of a
+# division and the second solver in a division as defined in secion 7.3.1 of
+# the SMT-COMP'19 rules.
+#
+# Note: The function prints a list of sorted tuples starting with the first
+#       place (winner).
+#
+def biggest_lead_ranking(data):
+    start = time.time() if g_args.show_timestamps else None
+
+    data = group_and_rank_solver(data)
+    data = data[data.competitive == True]
+    scores = []
+    # TODO: use for (year, division) groupby(level ...)
+    for year, ydata in data.groupby('year'):
+        for division, div_data in ydata.groupby('division'):
+
+            # Skip non-competitive divisions
+            if not is_division_competitive(div_data.solver.unique()):
+                continue
+
+            assert len(div_data) >= 2
+            first = div_data.iloc[0]
+            second = div_data.iloc[1]
+            score = ((1 + first.correct) / (1 + second.correct))
+            scores.append(
+                (score, first.solver, second.solver, division))
+
+    scores_sorted = sorted(scores, reverse=True)
+    print('Biggest Lead Ranking (Score, 1st Solver, 2nd Solver, Division)')
+    for s in scores_sorted:
+        print(s)
+
+    if g_args.show_timestamps:
+        log('time biggest_lead_ranking: {}'.format(time.time() - start))
+
+
+# Largest Contribution Ranking.
+#
+# Compute the correctly solved score for the virtual best solver for a given
+# division if the results of 'solver' are excluded. This function corresponds
+# to function vbss(D,S) as defined in section 7.3.2 of the SMT-COMP'19 rules.
+#
+def vbss(division_data, solver):
```

Comment:
We actually will want to normalize this with respect to the number of participants per division - else the comparison for the competition-wide recognition is not fair.

Reply:
@selig @mpreiner what do think of the following normalization:

Goal: Have "pieces of the pie" for two vs for `n` correspond to each other in relation, e.g., 50% in the pie for two is as much worth as 25% in the pie for 4. Then the formula would change as follows.

```
(1 - vbss(D, S-s) / vbss(D, S)) * n/2
```
where n is the number of solvers in a division.

Reply:
(The total would not sum up to 1 for n > 2, of course.)

Reply:
I've been thinking about this issue today and I still can't shake a gut feeling that we don't need normalisation. I understand the issue that without it the divisions where there are more participants are harder to get the 'largest' contribution in but I wouldn't be surprised if we see good contribution scores in those divisions even without normalisation e.g. there will be solvers that outperform other solvers on large parts of the division. I also don't see it as an intrinsically bad thing if less popular divisions benefit as there is some motivation to work on them. However, having said that, on balance I agree that we should explore a way of normalising.

With respect to your proposal, I was thinking along similar lines earlier but also realised that we wouldn't get numbers summing to 1 and worried that it would immediately bias the results towards more 'popular' divisions. But I don't know what we would see in practice. I guess we should test it out to see what numbers we get out. 

I thought I remembered that we were only looking at divisions with at least three competing solvers but we don't say that in the rules and perhaps that was something to do with the journal paper. 

Another thought is that, in general, it's a bit dangerous to base the scores on the number of participants as in the future people could 'game' the scores by adding extra variants etc. But this doesn't seem to be a big enough issue for us not to do this.

Reply:
Right, we did discuss excluding divisions with less than 3 participants in the past but somehow it didn't make it into the rules. The solver variants is also a good point, maybe they should actually somehow count as one (albeit not in the sense of counting all their contributions as one solver), but I don't have a good idea for how to deal with this.

As for the normalization question. The more I think about it, the more it seems to me that this is quite a difficult question. Even though I agree that encouraging people to work on less popular divisions is not necessarily a bad thing, it still feels like comparing apple and oranges if we don't try to treat every division equally.

Reply:
@selig @aniemetz What about normalizing based on the number of job pairs? For example, division `A` has `n_A` competitive solvers and `m_A` benchmarks and `N` is the overall number of job pairs (sum of `n_d * m_d` of all competitive divisions `d`). Then the weight for division `A` is `n_A*m_A/N`.

Reply:
So this would essentially give more weight to a division with 10 solvers and 3000 benchmarks over one with 2 solvers and 3 benchmarks -- which is actually very reasonable. The number of benchmarks is indeed an additionally angle to consider here, not only the number of solvers. This actually sounds good to me.